### ISSUE TITLE
Build hang issue

### ISSUE THREAD

**Issue by gi12345** [aws/aws-sdk-cpp#1887]
### Describe the issue

build / make is stuck when building target aws-cpp-sdk-ebs.
When building the SDK the make process always gets stuck on this target. Environment: 
Hardware: AWS-EC2 8 GB, Instance type t4g.large
Ubuntu Server 20.04 LTS, 64 bit ARM version
Log:
[ 55%] Building CXX object aws-cpp-sdk-ds/CMakeFiles/aws-cpp-sdk-ds.dir/ub_DS.cpp.o
[ 55%] Linking CXX shared library libaws-cpp-sdk-ds.so
[ 55%] Built target aws-cpp-sdk-ds
[ 56%] Building CXX object aws-cpp-sdk-dynamodb/CMakeFiles/aws-cpp-sdk-dynamodb.dir/ub_DYNAMODB.cpp.o
[ 56%] Linking CXX shared library libaws-cpp-sdk-dynamodb.so
[ 56%] Built target aws-cpp-sdk-dynamodb
[ 56%] Building CXX object aws-cpp-sdk-ebs/CMakeFiles/aws-cpp-sdk-ebs.dir/ub_EBS.cpp.o
[ 56%] Linking CXX shared library libaws-cpp-sdk-ebs.so
[ 56%] Built target aws-cpp-sdk-ebs
[ 56%] Building CXX object aws-cpp-sdk-ec2/CMakeFiles/aws-cpp-sdk-ec2.dir/ub_EC2.cpp.o
<==== Hangs here forever.

### Steps to Reproduce

check out AWS SDK 
cmake /home/ubuntu/awssdk_mar09/aws-sdk-cpp -DNO_HTTP_CLIENT=ON -BUILD_SHARED_LIBS
...........
-- TARGET_ARCH not specified; inferring host OS to be platform compilation target
-- Building AWS libraries as shared objects
-- Generating linux build config
-- Building project version: 1.9.211
--   Zlib library: /usr/lib/aarch64-linux-gnu/libz.so
-- Encryption: Openssl
--   Openssl include directory: /usr/include
--   Openssl library: /usr/lib/aarch64-linux-gnu/libssl.so;/usr/lib/aarch64-linux-gnu/libcrypto.so
-- You will need to inject an http client implementation before making any http requests!
-- AWS CRT C++ 1.9.211
-- CMake 3.16.3
-- Packaging currently only supported on Fedora.
-- Detected CMAKE_SYSTEM_PROCESSOR as aarch64
-- Detected 64-Bit system
............

make
[ 56%] Building CXX object aws-cpp-sdk-ec2/CMakeFiles/aws-cpp-sdk-ec2.dir/ub_EC2.cpp.o
<==== Hangs here forever.

### AWS CPP SDK version used

1.9.211

### compiler and version used

gcc

### Operating System and version

Ubuntu Server 20.04 LTS, 64 bit ARM version

**Comment by serzhick** [comment]
Hello, thank you for contacting us.

Unfortunately, some service clients began to be too big to be built on average systems (including ones with 8GB of ram) and we have [UNITY_BUILD](https://en.wikipedia.org/wiki/Unity_build) enabled by default
```
option(ENABLE_UNITY_BUILD "If enabled, the SDK will be built using a single unified .cpp file for each service library.  Reduces the size of static library binaries on Windows and Linux" ON)
```
If applicable to your case, I would suggest 
* disabling unity build by setting `-DENABLE_UNITY_BUILD=OFF`;
* reducing the number of build jobs to 1 (I think it is already your default case);
* getting a build machine with more RAM / enabling more virtual memory/swap (it will still take a lot of time due to IO/CPU limitations).

Best regards,
Sergey

**Comment by gi12345** [comment]
Thanks Sergey,
How much RAM, SWAP, ... is recommended? Do you know the corresponding AWS EC2 machine Instance Type I have to order to get the build done?

**Comment by jmklix** [comment]
I'm not sure how much ram you need to build all of the services, because you generally only want to build the ones that you need. You can do this by adding `-DBUILD_ONLY="s3;ec2"` to you cmake command. 

More guidance on building this sdk can be found [here](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-linux.html).

**Comment by serzhick** [comment]
I'm sorry, but we don't provide any recommendation for a build machine to use.
From a personal perspective, I was able to build the whole SDK in a default Debug configuration using c5.2xlarge instance (16Gb of RAM), please note that is a x86 instance, ARM may need different amount of resources. Please also note that by default ec2 instances do not provide swap memory, you may want to get more RAM or allocate swap area to avoid out-of-memory killer / machine stability issues.

**Comment by gi12345** [comment]
Thanks for your replies
I tried with parameter -DENABLE_UNITY_BUILD=OFF. Crashed (hangup) after more than 2 hours with just 8% completed. Worse than before, where 56% were completed in ~3hours (see initial log above).
---
I will go ahead and build just s3. I tried that before and run into issues on static / dynamic library creation and linkage. I can post the issues either here or create a new issue for it (please let me know).
Thanks

**Comment by serzhick** [comment]
If you see exactly the same / very similar log to the initial one above - it means that ENABLE_UNITY_BUILD is still ON, because I would expect you to see individual translation unit (basically .cpp files) being built. Could you please check that cmake cache is cleared before running cmake with a different config?
You could also enable parallel jobs run with -j 2 with unity build off with 2 core cpu / 8gb of ram because our individual translation units are relatively small.

Also, Joseph is right that building only individual service clients you need will be very beneficial for you having only 2 vCPUs on your build machine (although it is not an initial issue you are facing). Regarding the linking issues, most probably they are dependencies on CredentialProviders and STS client (as many service clients are using them), I would suggest this build config line `-DBUILD_ONLY="core;sts;identitystore"
Long story short for s3 and ec2, it still will take quite some time due to the amount of cpp files:
```
# clean up cmake cache
cmake /path/to/cpp/sdk -DENABLE_UNITY_BUILD=OFF -DBUILD_ONLY="core;sts;identitystore;s3;ec2"
cmake --build . -j 2
```
